src/app/shared/services/http-v2/http-v2.service.ts

```diff
@@ -0,0 +1,224 @@
+import { HttpClient, HttpHeaders, HttpParams } from '@angular/common/http';
+import { Injectable } from '@angular/core';
+import { StorageService } from '../storage/storage.service';
+import { environment } from '@root/environments/environment';
+import { Observable } from 'rxjs';
+import { map } from 'rxjs/operators';
+import { Location } from '@angular/common';
+
+const CSRF_KEY = 'CSRF';
+const AUTH_KEY = 'AUTH_TOKEN';
+
+type HttpMethod = 'post' | 'get' | 'put' | 'delete';
+type ResponseClass<T> = new (data: any) => T;
+
+interface RequestOptions {
+  csrf?: boolean;
+  authToken?: boolean;
+}
+
+const defaultOptions: RequestOptions = {
+  csrf: false,
+  authToken: true,
+};
+
+export function getFirst<T>(observable: Observable<T[]>): Observable<T> {
```

Comment:
This is very useful and it's a little surprising that we need it - do we return everything as an array?

Reply:
A tricky thing about designing all of this is that some endpoints will return an array of objects and some may only return a single object. In the "V1" code, we return objects that represent the returned `ResponseVO`. These objects will usually have separate methods to grab the data from the `ResponseVO` as either a single value or an array (something like `getArchiveVO()` vs `getArchiveVOs()`). Since Stela is returning data in a more plain format, there's not reason to package everything up in a "response object", since we don't need to write functionality to parse the response object for the data we actually need. But as mentioned earlier... some endpoints may return a single object of type `T` while others will return an array of type `T[]`.

The question then is: how do we want to deal with this? Ideally I don't want the caller to have to make a different functional call for getting a single object vs an array (having single and array versions of each HTTP method sounds like a lot), and I don't think there's a way to effectively do this through TypeScript's type magic. The closest we could get is having all of the `get`/`post`/`put`/`delete` methods return a type of `Observable<T | T[]>`... which ultimately always makes it the responsibility of the caller to do type guarding to determine whether they received an array of objects or a single object.

One of the goals of this service was to keep the typing of everything a bit more locked down and pleasant, so ultimately I decided that we should just *always* return type `T[]` so that everything's a bit more consistent. In the event that the API does return a single object, it will wrap it up into an array of length one. A consequence of this is that it is now slightly annoying if a method *is* fetching a single object (you always have to grab index 0 of the returned array), so I created this helper method to convert an `Observable<T[]>` to `Observable<T>`. That way, when this Observable is converted to a Promise, the Promise will only ever resolve with a single object instead of the array.

If anyone has any better ideas of how to handle this, please let me know!

(I also wish I could have extended the Observable class so calls could look like `this.httpv2.get(...).getFirst().toPromise(...)` instead of `getFirst(this.httpv2.get(...)).toPromise(...)`... but that would not really work)

Reply:
This is fascinating, thanks so much for the full explanation! I find it quite readable as is and it makes sense why you decided to do it this way. I don't think I have a better idea.